### ISSUE TITLE
Markdown Preview: ParseError: KaTeX parse error: Expected 'EOF', got '#'

### ISSUE THREAD

**Issue by rioj7** [microsoft/vscode#138970]
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.62.2
- OS Version: Windows 8.1

Steps to Reproduce:

1. Create a markdown file with the following content:
```
* `${dateTimeFormat}` : use the setting `templates.dateTimeFormat` to construct a [date-time](#variable-datetimeformat).
* <code>${dateTimeFormat:<em>name</em>:}</code> : use a _named_ format in the setting `templates.dateTimeFormat` to construct a [date-time](#variable-datetimeformat). The format properties override what is defined in `templates.dateTimeFormat`.
* `${input:description:}` : Ask the user some text and use the _`properties`_ part as the description for the InputBox<br/>Example: `${input:Title of this page:}`
* <code>${input:<em>description</em>:}</code> : Ask the user some text and use the _`properties`_ part as the description for the InputBox<br/>Example: 
* <code>${input:<em>description</em>:}</code> : Ask the user some text and use the _`properties`_ part as the description for the InputBox<br/>Example: `${input:Title of this page:}`
* <code>${snippet:<em>definition</em>:}</code> : you can use the full syntax of the [Visual Studio Code snippets](https://code.visualstudio.com/docs/editor/userdefinedsnippets#_snippet-syntax).<br/>A snippet is evaluated after the file is created with the command: **Next Snippet in File** from the Context Menu or Command Palette. The editor needs to be put in _Snippet_ mode. Apply this command for every `${snippet}` or `${cursor}` variable still in the file.<br/>
  Example: `${snippet##${1|*,**,***|} ${TM_FILENAME/(.*)/${1:/upcase}/} ${1}##}`
```
2.  Open a Preview to the side
3.  Rendering/Parsing of line 5 goes wrong
4.  Parsing of line 6 gives the: `ParseError: KaTeX parse error: Expected 'EOF', got '#'`
5.  I have added line 3 and 4 to determine what causes the problem, it is the literal text at the end of line 5: `${input:Title of this page:}`
6.  Any literal text (single backtick) of the content: `${xxx}` generates the error
7.  If there is a `<code>` tag on a line, any literal text (single backtick) after that is not recognized by the syntax highlighting in the editor

**Comment by rioj7** [comment]
I have just updated the extension [File Templates](https://marketplace.visualstudio.com/items?itemName=rioj7.vscode-file-templates) that has this piece of Markdown in the README.

Github, Marketplace and Extension Bar all render the markdown correct.

**Comment by mjbvz** [comment]
Minimal example that demonstrates the issue:

```md
<code>$x</code> `${a}`
```

**Comment by mjbvz** [comment]
Marking as a good first issue since this should be a fairly scoped bug

Here's the relevant repository where this issue should be fixed: https://github.com/mjbvz/markdown-it-katex

The repo includes tests that you can use to confirm the issue is fixed and also check for regressions

**Comment by JihongGan** [comment]
I am looking to fix this by not allowing a dangling dollar symbol inside a markdown block to open or close. Do you think I'm on the right path?

**Comment by Lemmingh** [comment]
I still do not regard the behavior here as a bug.

Things just work this way as per the inline parsing strategy of CommonMark.

---

From a CommonMark-compliant parser's view, we can say in a non-normative way that

```
<code>$x</code> `${a}`
```

is equivalent to

```
<xxxx>$x<xxxxx> `${x}`
```

Then, given the enabled rules (`markdown-it` default, CommonMark HTML, pandoc-style math, and something else), it's recognized as

```
<xxxx>$xxxxxxxxxx$xxxx
```

This is correct and exactly reflects the philosophy of CommonMark.

---

This issue ought to be marked as <https://github.com/microsoft/vscode/labels/%2Aas-designed>

**Markdown flavors are not compatible**, [as I mentioned above](https://github.com/microsoft/vscode/issues/138970#issuecomment-997447768). A derivative flavor can inherit the philosophy of its parent, but cannot retain the semantics.

In other words, you **cannot enable a math plugin and expect GFM-like rendering at the same time**. You can only discuss right or wrong within the same flavor.

If you want GFM-like rendering, disable the built-in Markdown Math (`vscode.markdown-math`) extension first.

**Comment by rioj7** [comment]
@Lemmingh But how do I write in Markdown Math a line that shows me

```
<code>$x</code> `${a}`
```

Is there a way to Escape the `$`

**Comment by Lemmingh** [comment]
```markdown
`$x` `${a}`
```

**Comment by Lemmingh** [comment]
Avoid plain HTML-like pieces in Markdown. Use proper syntax whenever possible.

In Markdown, HTML-like pieces (leaf block "HTML block" and inline structure "Raw HTML") are nothing, but just a kind of **literal** content. CommonMark requires implementations to **skip** them: There are roughly 7 + 6 rules for recognizing HTML-like pieces; HTML-like pieces are captured, and then emitted as-is to the output, without further touching parser's machine state.

**This design is a compromise for compatibility, and causes usability and security issues**, like your cases. **jgm** complained about the terrible fact many years ago.

---

As for escaping, if your use case is presenting a character in textual content, you can try "backslash escape" like `\$`, and "entity and character reference" like `&dollar;` and `&#36;`.

**Comment by rioj7** [comment]
@Lemmingh  Maybe the wrong example asked, the OP was about some _italic_ (or **bold**) text inside a code block, it was done with `<code>foo<em>bar</em>foo</code>`

<code>foo<em>bar</em>foo</code>

 and when you use a `$` character outside backticks and further down on the line a `$` it goes wrong.

I did not know Markdown allowed HTML entities.

I have now enabled **Markdown Math** and replaced all `$` characters inside `<code></code>` with `&dollar;` and now there is no `KaTeX parse error`

**Comment by Lemmingh** [comment]
Excuse me? I'm confused now.

So, you add these dangerous pieces to your Markdown document for formatting?

```markdown
<code>${input:<em>description</em>:}</code>

<code>foo<em>bar</em>foo</code>
```

Unfortunately, CommonMark Markdown is **not** a superset of WHATWG HTML. They actually have almost **no** intersection, and apply nearly **opposite** handling when encountering pieces that look like HTML tags.

If you decide to have HTML-like things in a Markdown document, you have to be careful, and know how to lead the parser to your desired result.

I thought I've made these basic facts clear. But seems no. No one understands my comments above.

---

Before I come up with another way to explain, we can demonstrate something by this sample:

```text
<section>
<!--

Hum?
-->
Oops!
</section>
```

If it's treated as HTML, there is one `section` element with a comment and a text node "`Oops!`" inside.

However, if you feed it to a CommonMark-compliant renderer, you'll get somewhat illegal HTML output:

```html
<section>
<!--
<p>Hum?
--&gt;
Oops!</p>
</section>
```

**Comment by rioj7** [comment]
@Lemmingh  If they are so dangerous why do they allow HTML tags in CommonMark Markdown?

As soon as you use single or triple backtick all inside is treated as literal, there is AFAIK no way to make something inside _italic_ (or **bold**) using Markdown syntax. I use HTML tags to get the required rendered output.

**Comment by Lemmingh** [comment]
> I use HTML tags to get the required rendered output

Good to know.

Then, you only need to be careful when configuring Markdown renderer and authoring Markdown documents. This is for not only performance but also **conformance**.

**Avoid using inappropriate parsers**, as flavors may hold dramatically different logic. For example, `cmark`, `Python-Markdown`, and `kramdown` can give mostly different results for the same input.

The Markdown flavor of GitHub is GFM. GitHub processes user content with [`cmark-gfm`](https://github.com/github/cmark-gfm), and then pipes the HTML output through some "filters" to add GitHub-specific features, such as sanitization, heading anchors, emoji codes, issue references, and diagrams.

[As I said above](https://github.com/microsoft/vscode/issues/138970#issuecomment-997447768), VS Code's `vscode.markdown-language-features` provides `markdown-it` default + `html: true`, which is a few steps away from GitHub. Basically, you just need to install extensions for emoji and task list, and disable those that don't exist on GitHub (e.g. `vscode.markdown-math`). Additionally, since VS Code and GitHub generate heading ID with slightly different methods, a link to heading might work on one and break on the other.

---

> If they are so dangerous why do they allow HTML tags in CommonMark

[For **compatibility**.](https://github.com/microsoft/vscode/issues/138970#issuecomment-1114468965)

It's John Gruber, [the primary inventor of Markdown](https://github.com/microsoft/vscode/issues/138970#issuecomment-997447768), that introduced "HTML" (but not real HTML). **The world has no choice but to live with it.**

If you ask me to judge Gruber's decision, I'd say the "HTML" in Markdown is a sign of laziness, besides, thousands of collective developer hours have been wasted due to tons of lazy and ambiguous design of the original Markdown. He should have defined generic construct syntax first, and built language structures on top of it.

---

Probably I'm bad at teaching. But other people have good skill in communication. I recommend their articles:

* [Beyond Markdown by **John MacFarlane**](https://talk.commonmark.org/t/2787)

* [markdownlint by **David Anson**](https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md)
